### ISSUE TITLE
Large discrepancy in gradients between jitted/non-jitted code

### ISSUE THREAD

**Issue by emilyfertig** [google/jax#12328]
### Description

The code below computes an element on the diagonal of a pairwise distance matrix, with a modified `sqrt` function that has finite gradients at 0. Without JIT, the gradients wrt inputs are 0., as expected. With JIT, the gradients are ~1e12.

This shows up in TFP-on-JAX's GaussianProcess. As a workaround, we can set the diagonal of the distance matrix to 0 (which zeros out the gradients), but that doesn't account for the case where the input array has duplicate points (so zeros can appear elsewhere in the distance matrix). This is blocking Vizier's use of TFP-on-JAX's GP.

```
import jax
from jax import numpy as jnp
import numpy as np

x = jnp.array(
    [-1.152883 , -1.229587 , -0.7161828,  1.6050261, -0.9966271, 1.0175147])

@jax.custom_gradient
def sqrt_with_finite_grads(x):
  """Avoid NaN gradients when `x == 0`."""
  def grad(grad_ys):
    large_float_like_x = np.sqrt(np.finfo(x.dtype).max)

    # In this repro, x == 0, so safe_grads = large_float_like_x.
    safe_grads = jnp.where(
        jnp.equal(x, 0.), large_float_like_x, 0.5 / jnp.sqrt(x))
    return grad_ys * safe_grads
  return jnp.sqrt(x), grad

def dist(x):
  """Stripped down from https://github.com/tensorflow/probability/blob/main/tensorflow_probability/python/math/psd_kernels/internal/util.py#L191"""
  # These two terms are equal. If we compute them differently (`sum(x*x)`, `dot(x, x)`, `sum(x**2)`) the issue persists. If we compute them the same way, the issue disappears.
  # (This is how we compute the diagonal of a squared distance matrix, when both inputs are the same.)
  row_norm_x = 2. * jnp.sum(x**2, axis=-1)
  middle_term = 2. * jnp.dot(x, x)

  # Their difference is 0.
  out = row_norm_x - middle_term
  d = jnp.clip(out, 0., jnp.inf)

  # Clipping at a very small value, instead of 0., appears to fix it. (Because `clip` zeros out the gradient when the value is outside the interval?)
  # d = jnp.clip(out, 1e-30, jnp.inf)

  return sqrt_with_finite_grads(d)

print('No jit grads: ', jax.grad(dist)(x))  # => [0., 0., ...]
print('Jit grads: ', jax.jit(jax.grad(dist))(x))  # => [-1.9e12, -1.7e12, ...]
```
[Colab](https://colab.sandbox.google.com/drive/1JSYnmaHWReXjYj55qsejDa8I-7lhPuCK#scrollTo=LfXHbn9vd3lb) with the same repro.

### What jax/jaxlib version are you using?

jax v0.3.17

### Which accelerator(s) are you using?

CPU

### Additional System Info

py3.7

**Comment by jakevdp** [comment]
Since you're depending on exact equality checks of floating point values, my initial suspicion is that this has to do with normal floating point accuracy differences between JIT and non-JIT code (discussed in e.g. [FAQ: JIT changes the behavior of my function](https://jax.readthedocs.io/en/latest/faq.html#jit-changes-the-behavior-of-my-function). In particular, the form of the computation you're using (factoring the euclidean distance into a square term and a cross term) is known to be quite unstable to floating point roundoff errors.

You can see this by running your code snippet at higher precision, with 64-bit computations enabled:
```
$ JAX_ENABLE_X64=true python script.py
```
With the following modification:
```python
print('X64:', jax.config.jax_enable_x64)
print('No jit grads: ', jax.grad(dist)(x))
print('Jit grads: ', jax.jit(jax.grad(dist))(x))
```
Here's the result with X64=false:
```
X64: False
No jit grads:  [0. 0. 0. 0. 0. 0.]
Jit grads:  [-1.8628299e+12 -1.6941562e+12 -6.2412056e+11  8.6855673e+11
 -7.4171023e+09  2.1605080e+12]
```
And X64=true
```
X64: True
No jit grads:  [0. 0. 0. 0. 0. 0.]
Jit grads:  [-1.87202911e-09 -2.41417555e-09  3.60670768e-09  2.58035980e-09
  1.88354525e-09  7.57478389e-10]
```
This strengthens the hypothesis that this is coming from floating point roundoff error in your functions.

I suspect the best approach would be to use a more numerically stable distance calculation; for example when I replace your `out` computation with this, it returns zero in both JIT and non-JIT cases:
```python
dist_matrix = x - x[:, None]) ** 2
out = dist_matrix.diagonal().sum()
```
Or, to avoid constructing the whole matrix, something like this:
```python
out = ((x - x) ** 2).sum()
```
In general, a computation where you are constructing two large floating point numbers and then subtracting them to get a value near zero is a bit of a numerical code smell, not just in JAX but in any computation based on IEEE floating point.

**Comment by emilyfertig** [comment]
I see, thanks so much for the quick/thorough response!

**Comment by emilyfertig** [comment]
Hi @jakevdp  -- I wanted to follow up because paring down the code substantially more, and removing the custom gradient/exact equality check, I'm still seeing the issue:

```
dist = lambda x: 1e6 * (jnp.sum(x**2, axis=-1) - jnp.dot(x, x))

x = jnp.array(
    [-1.152883 , -1.229587 , -0.7161828,  1.6050261, -0.9966271, 1.0175147])

print('No jit grads: ', jax.grad(dist)(x))  #  => [0. 0. 0. 0. 0. 0.]
print('Jit grads: ', jax.jit(jax.grad(dist))(x))  # => [-0.1056  0.0822 -0.0308  0.0018  -0.0597   0.0363]
```

I see your point about the numerical instability of the distance computation, but I was wondering if this looked to you like a large enough discrepancy that it's worth investigating further into the XLA compilation. Computing distance like this lets us avoid materializing prohibitively large intermediates in some cases, so it would be great if it worked under JAX/XLA (I understand if your answer's the same though). Thanks again.

**Comment by jakevdp** [comment]
Thanks for re-opening and for adding the concise reproduction! I think this is the same issue as above, and is behaving as expected.

You're essentially computing the same quantity using two different floating-point approaches, and subtracting them to attain a value of $0 \pm \varepsilon$ where float32 arithmetic means that the error should be approximately $\varepsilon \sim 10^{-7}$, due to float32 having 23-bits of mantissa. You then multiply that $\varepsilon$ by $10^6$, so we'd expect the result to be of order 0.1.

You can see this behavior in action without even using JAX; for example, here's a numpy computation where we use the identity $\sin^2 x + \cos^2 x = 1$ to compute a result two different ways, then subract them and multiply by a large scaling factor:
```python
import numpy as np

def f1(x):
  return np.sin(x) ** 2

def f2(x):
  return 1 - np.cos(x) ** 2

def f(x):
  return 1E6 * (f1(x) - f2(x))

np.random.seed(758493)
x = np.random.rand(6).astype('float32')
print(f(x))
# [-0.05960464 -0.06332994  0.05960464  0.         -0.05960464 -0.02980232]
```
Mathematically this result should be exactly zero, and it would be if floating point math were exact. But because floating point math is only an approximation of real math, and because we've set up the computation to exaggerate the small floating point error between the two computations, the result is not close to zero.

Back to your example: the fact that in non-JIT mode your code returns exactly zero is probably some accident of the implementation of the two approaches. Does that make sense?

**Comment by emilyfertig** [comment]
Got it, thanks so much for the additional explanation!
